tests/Unit/NumericalAlgorithms/Interpolation/Test_RegularGridInterpolant.py

```diff
@@ -0,0 +1,148 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import unittest
+from spectre import Spectral
+from spectre import DataStructures
+from spectre import Interpolation
+from numpy.polynomial.legendre import Legendre
+import numpy as np
+
+
+class TestRegularGrid(unittest.TestCase):
+
+    #random polynomial functions of low order for exact interpolation
+    def polynomial1D(self, points):
+        return points**2 + points + 1
+
+    def polynomial2D(self, points):
+        result = np.empty(len(points))
+        for i, p in enumerate(points):
+            result[i] = 2 * p[0]**2 + p[1]**2 + p[1] + p[0] + 2
+        return result
+
+    def polynomial3D(self, points):
+        result = np.empty(len(points))
+        for i, p in enumerate(points):
+            result[i] = 3 * p[0] ** 2 + 2 * p[1] ** 2 + p[2] ** 2 + 2 * p[1] + \
+                        p[2] + p[0] + 2
+        return result
+
+    def generate_Gauss_nodes(self, num_points):
+        return Legendre.basis(num_points).roots()
+
+    def generate_GaussLobatto_nodes(self, num_points):
+        nodes = Legendre.basis(num_points - 1).deriv().roots()
+        return np.concatenate(([-1], nodes, [1]))
+
+    def generate_mesh(self, nodes, dim):
+        """
+        creates a uniform mesh from the provided nodes
+        """
+        return np.stack(np.meshgrid(*tuple(nodes for _ in range(dim))),
+                        -1).reshape(-1, dim)
+
+    def check_interpolation(self, interpolant, initial_points, target_points,
+                            polynomial):
+        initial_data = polynomial(initial_points)
+        target_data = polynomial(target_points)
+        interpolated_data = interpolant.interpolate(
+            DataStructures.DataVector(initial_data))
```

Comment:
I don't quite understand why this works, or does it actually? If I understand correctly `initial_data` is a dim-dimension array, at least that's what the `polynomial` function returns, but `DataVector` expects a flattened array. I suggest you make clearer what the shapes of arrays are here. How about:
- Always work with `source_coords` and `target_coords` that are shaped `(dim, num_points)`, i.e. each component is flattened over the grid, but the components are not flattened.
- The polynomials return scalars. (I don't really understand how their components are currently handled)
- Name `generate_mesh` -> `logical_coordinates(mesh)` and put all the logic to generate the `(dim, num_points)`-shaped coords into that function. Then you don't need to have an extra test case per quadrature case below, but you can loop over the enum cases. If you prefer, you could also bind the `logical_coordinates(mesh)` function in `src/Domain/LogicalCoordinates.hpp`, which does the same thing. It returns a `tnsr::I` in C++, but you could make it return a numpy array in Python instead.